File: agents/android/src/net/enroll.ts
/**
 * One-time enrollment, ported from agents/windows/internal/enroll/enroll.go.
 *
 * NOTE the casing split: this REST body is snake_case, unlike everything on the
 * WebSocket, which is camelCase. The response is camelCase.
 */
import {deriveWsUrl, Identity} from '../config/storage';
import {getDeviceTelemetry} from '../telemetry';
import {log} from '../log';

interface EnrollResponse {
  deviceId: string;
  agentSecret: string;
  wsUrl: string;
  heartbeatIntervalSeconds: number;
}

/**
 * Redeem an enrollment token and return the permanent identity. The caller persists it.
 * The old stub POSTed correctly but threw the response away — this keeps it.
 */
export async function enroll(
  serverUrl: string,
  token: string,
  name: string,
): Promise<Identity> {
  const base = serverUrl.trim().replace(/\/+$/, '');
  const telemetry = await getDeviceTelemetry();
  log(`enrolling at ${base} as "${name}" (${telemetry.osVersion})`);

  const res = await fetch(`${base}/api/devices/enroll`, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({
      enrollment_token: token.trim(),
      name: name.trim(),
      type: 'android',
      os_version: telemetry.osVersion,
    }),
  });

  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(
      res.status === 400 || res.status === 404
        ? 'Invalid or expired enrollment token.'
        : `Enrollment failed (${res.status}). ${text}`.trim(),
    );
  }

  const body = (await res.json()) as EnrollResponse;
  if (!body.deviceId || !body.agentSecret) {
    throw new Error('Enrollment response missing device identity.');
  }
  log(`enrolled: deviceId=${body.deviceId.slice(0, 8)}… wsUrl=${body.wsUrl}`);

  return {
    serverUrl: base,
    wsUrl: body.wsUrl || deriveWsUrl(base),
    deviceId: body.deviceId,
    agentSecret: body.agentSecret,
    heartbeatIntervalSeconds: body.heartbeatIntervalSeconds || 10,
  };
}

File: agents/windows/internal/enroll/enroll.go
// Package enroll exchanges a one-time enrollment token for this device's permanent
// identity.
package enroll

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"runtime"
	"strings"
	"time"

	"drs/agent/windows/internal/config"
	"drs/agent/windows/internal/sysinfo"
)

type enrollRequest struct {
	EnrollmentToken string `json:"enrollment_token"`
	Name            string `json:"name"`
	Type            string `json:"type"`
	OSVersion       string `json:"os_version"`
}

type enrollResponse struct {
	DeviceID                 string `json:"deviceId"`
	AgentSecret              string `json:"agentSecret"`
	WSURL                    string `json:"wsUrl"`
	HeartbeatIntervalSeconds int    `json:"heartbeatIntervalSeconds"`
}

type errorResponse struct {
	Error string `json:"error"`
}

// Enroll redeems a token and returns the identity to persist. deviceName may be empty,
// in which case the machine's hostname is used.
func Enroll(serverURL, token, deviceName string) (config.Config, error) {
	serverURL = strings.TrimRight(serverURL, "/")
	if deviceName == "" {
		if h, err := os.Hostname(); err == nil {
			deviceName = h
		} else {
			deviceName = "Unnamed device"
		}
	}

	body, err := json.Marshal(enrollRequest{
		EnrollmentToken: token,
		Name:            deviceName,
		Type:            runtime.GOOS,
		OSVersion:       sysinfo.Info().OS,
	})
	if err != nil {
		return config.Config{}, err
	}

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Post(serverURL+"/api/devices/enroll", "application/json", bytes.NewReader(body))
	// Check err before touching resp. The previous agent read resp.StatusCode in the
	// same branch as a non-nil error, which panicked on a nil resp every time the
	// server was unreachable.
	if err != nil {
		return config.Config{}, fmt.Errorf("could not reach %s: %w", serverURL, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		var er errorResponse
		_ = json.NewDecoder(resp.Body).Decode(&er)
		if er.Error != "" {
			return config.Config{}, fmt.Errorf("enrollment refused (%d): %s", resp.StatusCode, er.Error)
		}
		return config.Config{}, fmt.Errorf("enrollment refused with status %d", resp.StatusCode)
	}

	var out enrollResponse
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return config.Config{}, fmt.Errorf("could not read enrollment response: %w", err)
	}
	if out.DeviceID == "" || out.AgentSecret == "" {
		return config.Config{}, fmt.Errorf("enrollment response was missing the device identity")
	}

	cfg := config.Config{
		ServerURL:                serverURL,
		WSURL:                    out.WSURL,
		DeviceID:                 out.DeviceID,
		AgentSecret:              out.AgentSecret,
		HeartbeatIntervalSeconds: out.HeartbeatIntervalSeconds,
	}
	if cfg.WSURL == "" {
		cfg.WSURL = config.DeriveWSURL(serverURL)
	}
	if cfg.HeartbeatIntervalSeconds <= 0 {
		cfg.HeartbeatIntervalSeconds = 10
	}
	return cfg, nil
}
